transports/tcp/src/lib.rs

```diff
@@ -365,30 +367,45 @@ where
         tracing::debug!(address=%socket_addr, "dialing address");
 
         let socket = self
+            .config
             .create_socket(socket_addr, opts.port_use)
             .map_err(TransportError::Other)?;
 
+        let mut bound_to_local = false;
+
         match self.port_reuse.local_dial_addr(&socket_addr.ip()) {
             Some(socket_addr) if opts.port_use == PortUse::Reuse => {
                 tracing::trace!(address=%addr, "Binding dial socket to listen socket address");
                 socket
                     .bind(&socket_addr.into())
                     .map_err(TransportError::Other)?;
```

Comment:
Given the comment on the `async` block below, I think we should also do the `bind` within the returned future. We can do that by storing the `Option<SocketAddr>`, then you don't need to keep the additional boolean.